File: sdk/dart/dovahlink_client/lib/src/protocol/pairing_outcome_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';
import 'package:dovahlink_client_sdk/src/protocol/pairing_outcome_payload_validator.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_format_exception.dart';
import 'package:dovahlink_client_sdk/src/shared/enums.dart';

part 'pairing_outcome_payload.g.dart';

/// Incoming `pairing_outcome` payload (`protocol/schema/README.md`'s `pairing_outcome`).
/// Decode-only: the client never sends `pairing_outcome`.
@JsonSerializable(checked: true, createToJson: false)
class PairingOutcomePayload {
  /// Creates a pairing-outcome payload from decoded Host values.
  /// [outcome] is the Host-reported result.
  /// [credential], [shortId], and [displayName] carry trust metadata when applicable.
  /// [attemptsRemaining] and [retryAfterSeconds] carry Host runtime metadata when applicable.
  const PairingOutcomePayload({
    required this.outcome,
    required this.credential,
    required this.shortId,
    required this.displayName,
    required this.attemptsRemaining,
    this.retryAfterSeconds,
  });

  /// Decodes and validates one `pairing_outcome` payload from [json].
  /// Returns the typed Host reply.
  /// Throws [ProtocolFormatException] if its shape or metadata is malformed.
  factory PairingOutcomePayload.fromJson(JsonMap json) {
    try {
      final PairingOutcomePayload payload = _$PairingOutcomePayloadFromJson(
        json,
      );
      PairingOutcomePayloadValidator.validate(
        outcome: payload.outcome,
        credential: payload.credential,
        shortId: payload.shortId,
        displayName: payload.displayName,
        attemptsRemaining: payload.attemptsRemaining,
        retryAfterSeconds: payload.retryAfterSeconds,
        json: json,
      );
      return payload;
    } on Object catch (error) {
      throw ProtocolFormatException('Invalid pairing_outcome payload: $error');
    }
  }

  /// The wire vocabulary of `pairing_outcome.outcome`, spanning every exchange it can reply to.
  @JsonKey(required: true)
  final PairingOutcome outcome;

  /// Present only for `"credential_issued"`, `"trusted"`, and `"already_trusted"`.
  @JsonKey(required: true)
  final String? credential;

  /// The administration-only identifier, present only for `"trusted"` and `"already_trusted"`.
  @JsonKey(required: true)
  final String? shortId;

  /// Echoes the client-supplied label, present only alongside [credential]/[shortId] when the
  /// client supplied one.
  @JsonKey(required: true)
  final String? displayName;

  /// The Host-calculated number of wrong-code attempts remaining after a counted `invalid` result;
  /// `null` when the attempt was not counted or the outcome is not `invalid`.
  @JsonKey(required: true)
  final int? attemptsRemaining;

  /// Host-authoritative whole seconds until retry is safe, present for `"pacing_limited"`,
  /// `"renotify_cooldown"`, and the newly committed cooldown on `"renotified"`; `null` otherwise.
  @JsonKey(required: true)
  final int? retryAfterSeconds;
}

File: integration/DovahLinkValidationClient/PairingOutcomePayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Incoming <c>pairing_outcome</c> payload (<c>protocol/schema/README.md</c>'s
/// <c>pairing_outcome</c>). Decode-only: this client never sends <c>pairing_outcome</c>. Shared
/// reply to <c>pairing_confirm</c>, <c>pairing_ack</c>, <c>pairing_renotify</c>, and
/// <c>pairing_cancel</c>, distinguished by <see cref="Outcome"/>.
/// </summary>
/// <param name="Outcome">The wire vocabulary of <c>pairing_outcome.outcome</c>.</param>
/// <param name="Credential">Present only for <c>"credential_issued"</c>, <c>"trusted"</c>, and
/// <c>"already_trusted"</c>.</param>
/// <param name="ShortId">The administration-only identifier, present only for <c>"trusted"</c> and
/// <c>"already_trusted"</c>.</param>
/// <param name="DisplayName">Echoes the client-supplied label, present only alongside
/// <paramref name="Credential"/>/<paramref name="ShortId"/> when the client supplied one.</param>
/// <param name="RetryAfterSeconds">The minimum safe wait in seconds before retrying, present for
/// <c>"pacing_limited"</c> and <c>"renotify_cooldown"</c>.</param>
public sealed record PairingOutcomePayload(
    string Outcome,
    string? Credential,
    string? ShortId,
    string? DisplayName,
    int? RetryAfterSeconds)
{
    /// <summary>
    /// Decodes and validates one <c>pairing_outcome</c> payload.
    /// </summary>
    /// <param name="payload">The envelope's decoded payload object.</param>
    /// <returns>The decoded pairing-outcome payload.</returns>
    /// <exception cref="FormatException">Thrown when a required key is missing or a value has the
    /// wrong JSON type.</exception>
    public static PairingOutcomePayload Decode(JsonObject payload)
    {
        try
        {
            string outcome = payload["outcome"]?.GetValue<string>() ?? throw new FormatException("Missing outcome.");
            return new PairingOutcomePayload(
                outcome,
                Credential: GetRequiredNullableString(payload, "credential"),
                ShortId: GetRequiredNullableString(payload, "shortId"),
                DisplayName: GetRequiredNullableString(payload, "displayName"),
                RetryAfterSeconds: GetRequiredNullableInt(payload, "retryAfterSeconds"));
        }
        catch (InvalidOperationException ex)
        {
            throw new FormatException($"Malformed pairing_outcome payload: {ex.Message}", ex);
        }
    }

    /// <summary>Extracts a required, nullable string field, rejecting an absent key.</summary>
    /// <param name="payload">The payload object.</param>
    /// <param name="key">The field's key.</param>
    /// <returns>The field's string value, or <see langword="null"/>.</returns>
    /// <exception cref="FormatException">Thrown when the key is absent.</exception>
    private static string? GetRequiredNullableString(JsonObject payload, string key)
    {
        if (!payload.ContainsKey(key))
        {
            throw new FormatException($"Missing {key}.");
        }
        return payload[key]?.GetValue<string>();
    }

    /// <summary>Extracts a required, nullable int field, rejecting an absent key.</summary>
    /// <param name="payload">The payload object.</param>
    /// <param name="key">The field's key.</param>
    /// <returns>The field's int value, or <see langword="null"/>.</returns>
    /// <exception cref="FormatException">Thrown when the key is absent.</exception>
    private static int? GetRequiredNullableInt(JsonObject payload, string key)
    {
        if (!payload.ContainsKey(key))
        {
            throw new FormatException($"Missing {key}.");
        }
        return payload[key]?.GetValue<int>();
    }
}
